h2o-algos/src/main/java/hex/tree/isoforextended/ExtendedIsolationForest.java

```diff
@@ -0,0 +1,378 @@
+            if (_parms.extension_level < 0 || _parms.extension_level > extensionLevelMax) {
+                throw new IllegalStateException("Parameter extension_level must be in interval [0, "
+                        + extensionLevelMax + "] but it is " + _parms.extension_level);
+            }
+            long sampleSizeMax = _parms.train().numRows();
+            if (_parms._sample_size < 0 || _parms._sample_size > sampleSizeMax) {
+                throw new IllegalStateException("Parameter sample_size must be in interval [0, "
+                        + sampleSizeMax + "] but it is " + _parms._sample_size);
+            }
+        }
+    }
+
+    @Override
+    protected double score1(Chunk[] chks, double offset, double weight, double[] fs, int row) {
+        // not used in the current implementation of Extended Isolation Forest
+        return 0;
+    }
+
+    @Override
+    protected Driver trainModelImpl() {
+        return new ExtendedIsolationForestDriver();
+    }
+
+    @Override
+    public ModelCategory[] can_build() {
+        return new ModelCategory[]{
+                ModelCategory.AnomalyDetection
+        };
+    }
+
+    @Override
+    protected ScoreKeeper.ProblemType getProblemType() {
+        return ScoreKeeper.ProblemType.anomaly_detection;
+    }
+
+    @Override
+    public boolean scoreZeroTrees() {
+        return false;
+    }
+
+    @Override
+    public boolean isSupervised() {
+        return false;
+    }
+
+    @Override
+    public boolean havePojo() {
+        return false;
+    }
+
+    @Override
+    public boolean haveMojo() {
+        return true;
+    }
+
+    @Override
+    protected void addCustomInfo(ExtendedIsolationForestModel.ExtendedIsolationForestOutput out) {
+        if (_iTrees != null) {
+            out.iTrees = _iTrees;
+        }
+    }
+
+    private class ExtendedIsolationForestDriver extends Driver {
+
+        @Override
+        protected ExtendedIsolationForestModel makeModel(Key<ExtendedIsolationForestModel> modelKey,
+                                                         ExtendedIsolationForestModel.ExtendedIsolationForestParameters parms) {
+            return new ExtendedIsolationForestModel(modelKey, parms, new ExtendedIsolationForestModel.ExtendedIsolationForestOutput(ExtendedIsolationForest.this));
+        }
+
+        @Override
+        protected boolean doOOBScoring() {
+            return true;
+        }
+
+        @Override
+        protected void scoreAndBuildTrees(boolean oob) {
+            IsolationTreeForkJoinTask [] iTreeTasks = new IsolationTreeForkJoinTask[_parms._ntrees];
+            for (int t = 0; t < _parms._ntrees; t++) {
+                iTreeTasks[t] = new IsolationTreeForkJoinTask();
+                H2O.submitTask(iTreeTasks[t]);
+            }
+            for (int t = 0; t < _parms._ntrees; t++) {
+                _iTrees[t] = iTreeTasks[t].getResult();
+                _job.update(1);
```

Comment:
This will work in most cases but it is not 100% correct. If for whatever reason the first tree takes much longer than others the progress bar will not be smooth, it will have jumps. It would be better if the IsolationTreeForkJoinTask updated the job progress as the tree is build.